Statically build and link the ANTLR4 runtime.

HackerFoo opened the pull request:
Statically link the ANTLR4 runtime to remove the runtime dependency. The source build will automatically pull necessary code from GitHub.

```
$ ldd build/lib.linux-x86_64-3.7/fasm/parser/libparse_fasm.so
	linux-vdso.so.1 (0x00007fff4cfb6000)
	libstdc++.so.6 => /nix/store/danv012gh0aakh8xnk2b35vahklz72mk-gcc-9.2.0-lib/lib/libstdc++.so.6 (0x00007f69947fb000)
	libm.so.6 => /nix/store/xg6ilb9g9zhi2zg1dpi4zcp288rhnvns-glibc-2.30/lib/libm.so.6 (0x00007f69946bb000)
	libgcc_s.so.1 => /nix/store/xg6ilb9g9zhi2zg1dpi4zcp288rhnvns-glibc-2.30/lib/libgcc_s.so.1 (0x00007f69946a1000)
	libc.so.6 => /nix/store/xg6ilb9g9zhi2zg1dpi4zcp288rhnvns-glibc-2.30/lib/libc.so.6 (0x00007f69944e2000)
	/nix/store/xg6ilb9g9zhi2zg1dpi4zcp288rhnvns-glibc-2.30/lib64/ld-linux-x86-64.so.2 (0x00007f6994b0e000)
```

HackerFoo commented:
This should be a much better solution than https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1880

litghost commented on third_party/antlr4/runtime/Cpp/cmake/ExternalAntlr4Cpp.cmake:
```diff
@@ -51,7 +51,7 @@ endif()
   # This avoids
   # 'warning: jobserver unavailable: using -j1. Add '+' to parent make rule.'
-  set(ANTLR4_BUILD_COMMAND $(MAKE))
+  set(ANTLR4_BUILD_COMMAND $(MAKE) CXX_FLAGS="-fPIC" C_FLAGS="-fPIC")
```
Changes in these files feels wrong.  Should these go upstream first?  I feel like something is missing.

HackerFoo replied:
This is a slightly strange way to build this, because I want to statically link one library to another shared library to produce a single shared library. The right way is to probably modify this file since this is a unique situation.

It would be more common to either package the shared libraries, or statically link them into an executable. So I don't expect this would get merged upstream.

HackerFoo replied:
I could copy this somewhere outside `third_party/` if that's preferred.

litghost replied:
I believe if you proposed something akin to the option `ANTLR4_WITH_STATIC_CRT` (see line 74-75) about enabling PIC even when building a static library, it would at a minimum start a valuable conversation with the ANTLR4 team and potentially be upstreamed.  I think that approach will pull in insight from the ANTLR4 folks, and lead to a better solution.

HackerFoo replied:
This will take time. I've created [an issue](https://github.com/SymbiFlow/fasm/issues/42), but I'd like to merge this for now.

litghost replied:
I think taking that time now is more valuable than landing this PR immediately.

HackerFoo replied:
@mithro What are your thoughts on this? I'd like to get this this week.

litghost replied:
This issue appears to have occurred in at least two other instances with ANTLR4:
 - https://github.com/antlr/antlr4/issues/2688
 - https://github.com/antlr/antlr4/issues/2776

Rather than maintain a local fix, we should get the fix in upstream.

HackerFoo replied:
I agree, but I don't think it needs block the work we've done on FASM being used today, so I've filed an issue.

HackerFoo replied:
I found a simple solution that doesn't require changing these files - just set CFLAGS/CXXFLAGS to `-fPIC`.

litghost reviewed (commented):
Would it be possible to make whether to compile statically or dynamically an option?  I can see a case for both configurations.

HackerFoo commented:
> Would it be possible to make whether to compile statically or dynamically an option? I can see a case for both configurations.

I'm sure it's possible, but it would take time. How useful is this feature? Could it be done in a later PR?

litghost commented:
> > Would it be possible to make whether to compile statically or dynamically an option? I can see a case for both configurations.
> 
> I'm sure it's possible, but it would take time. How useful is this feature? 

We've already gotten requests in prjxray to enable unvendored dependencies, see https://github.com/SymbiFlow/prjxray/issues/1376 .  While I'm never excited about adding work for packaging, I'm also unexcited about knowingly creating work for the future when it can be avoided.

> Could it be done in a later PR?

I don't believe so.

HackerFoo commented:
What would be the advantage of being able to select a shared library in this specific case?

> > > Would it be possible to make whether to compile statically or dynamically an option? I can see a case for both configurations.
> > 
> > 
> > I'm sure it's possible, but it would take time. How useful is this feature?
> 
> We've already gotten requests in prjxray to enable unvendored dependencies, see [SymbiFlow/prjxray#1376](https://github.com/SymbiFlow/prjxray/issues/1376) . While I'm never excited about adding work for packaging, I'm also unexcited about knowingly creating work for the future when it can be avoided.
> 
> > Could it be done in a later PR?
> 
> I don't believe so.

What would be the advantage of being able to select a shared library in this specific case? The ANTLR runtime is not a large dependency. Note that this doesn't include the ANTLR parser generator itself, which is only used at compile time.

HackerFoo commented:
I have added a flag to use an external shared library for the ANTLR runtime:

```
$ python setup.py build_ext --help
Common commands: (see '--help-commands' for more)
...

Options for 'CMakeBuild' command:
  --antlr-runtime  Whether to use a 'static' or 'shared' ANTLR runtime.
  --help-compiler  list available compilers

...
```

`--antlr-runtime=static` is the default, and `--antlr-runtime=shared` will look for `libantlr-runtime` using CMake's `find_library`

litghost commented on setup.py:
```diff
@@ -105,6 +114,16 @@ def build_extension(self, ext):
+        super().finalize_options()
+        assert self.antlr_runtime in [
+            'static', 'shared'
+        ], 'Invalid antlr_runtime'
```
Change message to `'Invalid antr_runtime {}, expected one of {}'.format(self.antr_runtime, ANTLR_RUNTIMES)`, etc?

HackerFoo replied:
Done.

litghost commented on tox.ini:
```diff
@@ -33,3 +33,4 @@ per-file-ignores =
 ignore =
     fasm/parser/tags.py
     fasm/parser/libparse_fasm.*
+    third_party/antlr4/runtime/PHP
```
Should probably ignore `third_party/antlr4` completely?

HackerFoo replied:
I'll add a comment here, but the reason this is needed specifically is because this is an empty directory, and this causes problems: https://stackoverflow.com/questions/2640378/include-empty-directory-with-python-setup-py-sdist

The typical workaround is to just put a file in there, but this is in a submodule.

HackerFoo replied:
Putting `third_party` here didn't work either.

litghost replied:
That seems fine.  Add a comment explaining that.

HackerFoo replied:
Done.

litghost commented on setup.py:
```diff
@@ -92,6 +99,8 @@ def build_extension(self, ext):
                 env.get('CXXFLAGS', ''), self.distribution.get_version())
             if not os.path.exists(self.build_temp):
                 os.makedirs(self.build_temp)
+            os.environ["CXXFLAGS"] = "-fPIC"
```
This looks good, maybe add a comment about why this is here?

litghost replied:
Worth doing the equivalent of `CXXFLAGS += -fPIC`?

HackerFoo replied:
Good idea.

HackerFoo replied:
Added a comment, and included previous flags from the environment.

litghost reviewed (commented):
Can GH actions add a matrix option on static/dynamic easily?

HackerFoo commented:
@litghost 

- I've removed the modified CMake files and added an unmodified submodule at `third_party/antlr4`. This avoids any maintenance of upstream code and possible divergence.
- I've added a flag to `setup.py` to allow using the system provided shared library for the ANTLR4 runtime, which will work better with modern OSes which provide this library.

litghost commented:
> @litghost
> 
>     * I've removed the modified CMake files and added an unmodified submodule at `third_party/antlr4`. This avoids any maintenance of upstream code and possible divergence.
> 
>     * I've added a flag to `setup.py` to allow using the system provided shared library for the ANTLR4 runtime, which will work better with modern OSes which provide this library.

I think this PR is looking better.  Just a couple small comments, and I think we are good to go.

HackerFoo commented:
> Can GH actions add a matrix option on static/dynamic easily?

Yes, I can do this. I will need to build the shared library as well.

litghost commented:
> > Can GH actions add a matrix option on static/dynamic easily?
> 
> Yes, I can do this. I will need to build the shared library as well.

I might be confused, but can't the `shared` library option just use the system version?  That's what the CI did before this PR, yes?

HackerFoo commented:
> > > Can GH actions add a matrix option on static/dynamic easily?
> > 
> > 
> > Yes, I can do this. I will need to build the shared library as well.
> 
> I might be confused, but can't the `shared` library option just use the system version? That's what the CI did before this PR, yes?

Yes it can, but there is no binary package available on CI, where the test will be run.

litghost commented:
> > > > Can GH actions add a matrix option on static/dynamic easily?
> > > 
> > > 
> > > Yes, I can do this. I will need to build the shared library as well.
> > 
> > 
> > I might be confused, but can't the `shared` library option just use the system version? That's what the CI did before this PR, yes?
> 
> Yes it can, but there is no binary package available on CI, where the test will be run.

Sorry, what am I missing?  This line is installing the OS package, yes?

https://github.com/SymbiFlow/fasm/blob/0448187babe4ec2408e8c821fed60401848b0a58/.github/workflows/presubmit.yml#L60

HackerFoo commented:
Right. I forgot, this is running on 20.04. Well, I've added the fallback now, anyway. I'm not sure how useful it is.

litghost reviewed (changes requested):
The current flag approach doesn't appear to work. As a concrete example, how (via pip) can a user install the `fasm` package using a system dynamic library instead of the static library?

HackerFoo commented:
> The current flag approach doesn't appear to work. As a concrete example, how (via pip) can a user install the `fasm` package using a system dynamic library instead of the static library?

I have added this flag to all the relevant commands, and documentation on how to use this with `pip` in the README.

litghost commented:
> > The current flag approach doesn't appear to work. As a concrete example, how (via pip) can a user install the `fasm` package using a system dynamic library instead of the static library?
> 
> I have added this flag to all the relevant commands, and documentation on how to use this with `pip` in the README.

I don't believe this works?  Maybe I'm missing something?

```
python3 -mvenv test_pip
source test_pip/bin/activate
pip install cython textx
pip install -v git+https://github.com/HackerFoo/fasm.git@static_antlr4_runtime --install-option="--antlr-runtime=shared"
ldd /usr/local/google/home/keithrothman/cat_x/test_pip/lib/python3.8/site-packages/fasm//parser/libparse_fasm.so 
```

Outputs:
```
$ ldd /usr/local/google/home/keithrothman/cat_x/test_pip/lib/python3.8/site-packages/fasm//parser/libparse_fasm.so 
        linux-vdso.so.1 (0x00007ffd0d976000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f590dad9000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f590d995000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f590d97b000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f590d7b6000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f590ddfe000)

```

For clarity, the following does work:
```
python3 setup.py build_ext --antlr-runtime=shared
ldd ./build/lib.linux-x86_64-3.8/fasm/parser/libparse_fasm.so
```
outputs:
```
$ ldd ./build/lib.linux-x86_64-3.8/fasm/parser/libparse_fasm.so
        linux-vdso.so.1 (0x00007ffebdff4000)
        libantlr4-runtime.so.4.8 => /usr/lib/x86_64-linux-gnu/libantlr4-runtime.so.4.8 (0x00007f9f72e6c000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f9f72c9f000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f9f72b5b000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f9f72b41000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f9f7297c000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f9f72ff4000)
```

litghost commented:
FYI, I know that whatever PyYAML is doing works with this: https://github.com/yaml/pyyaml/blob/master/setup.py#L120

PyYAML has an option to use libyaml, with a python fallback (kind of like what we are doing).  I debugged PyYAML not using the binary package with the following invocation:
```
pip install  --global-option='--with-libyaml' pyyaml
```

HackerFoo commented:
@litghost It looks like a known bug with VCS urls: https://github.com/pypa/pip/issues/5518

litghost commented:
> @litghost It looks like a known bug with VCS urls: [pypa/pip#5518](https://github.com/pypa/pip/issues/5518)

Fixed in 2019? https://github.com/pypa/pip/pull/6389 ?

litghost commented:
> @litghost It looks like a known bug with VCS urls: [pypa/pip#5518](https://github.com/pypa/pip/issues/5518)

I am testing whether the install from local directory works and the VCS does not.

HackerFoo commented:
Even when I try random flags like `--with-foo=bar`, nothing happens, so I'm pretty sure flags are being ignore entirely. I'm trying to figure out why.

HackerFoo commented:
The flag seems to affect pyyaml's setup.py even when using a Git URL, so that isn't it.

HackerFoo commented:
The presence of `pyproject.toml` seems to cause flags to be ignored. It works when I remove the file.

HackerFoo commented:
@litghost I found that the `--no-use-pep517` flag for `pip` will allow flags to work as expected. This is needed because the existence of `pyproject.toml` enables pip's PEP517 mode.

litghost commented on tox.ini:
```diff
@@ -17,6 +17,7 @@ commands =
     # This repository uses a Markdown long_description, so the -r flag to
     # `setup.py check` is not needed. If your project contains a README.rst,
     # use `python setup.py check -m -r -s` instead.
+    python setup.py build_ext --antlr-runtime={env:ANTLR4_RUNTIME_TYPE:static}
```
I think this should be changed to `pip install -e . --antlr-runtime={env:ANTLR4_RUNTIME_TYPE:static} --no-use-pep517`?

HackerFoo replied:
Done.

litghost commented:
> @litghost I found that the `--no-use-pep517` flag for `pip` will allow flags to work as expected. This is needed because the existence of `pyproject.toml` enables pip's PEP517 mode.

Weird!  Good find.

litghost commented on tox.ini:
```diff
@@ -17,6 +17,7 @@ commands =
     # `setup.py check` is not needed. If your project contains a README.rst,
     # use `python setup.py check -m -r -s` instead.
+    pip install -e . --install-option="--antlr-runtime={env:ANTLR4_RUNTIME_TYPE:static}" --no-use-pep517
     python setup.py develop
```
I believe the `setup.py develop` isn't needed (implied by the `-e) and the check isn't needed either?

litghost replied:
Also, if the `-e` is breaking things, I think we can drop it IMO.

HackerFoo replied:
I've removed these lines. `-e` works fine.

litghost reviewed (commented):
LGTM, the `--no-pep-517` does appear to fix the problem with the flags.  Can you add a comment in both places you mention/use the `--no-pep-517` flag with why it is there.  Relevant issue: https://github.com/pypa/pip/issues/5771

HackerFoo commented:
> LGTM, the `--no-pep-517` does appear to fix the problem with the flags. Can you add a comment in both places you mention/use the `--no-pep-517` flag with why it is there. Relevant issue: [pypa/pip#5771](https://github.com/pypa/pip/issues/5771)

Done, and I've added an issue to update when there is a better way to do this: https://github.com/SymbiFlow/fasm/issues/44
